pkg/mcp/server/monitoring.go

```diff
@@ -0,0 +1,202 @@
+// limitations under the License.
+
+package server
+
+import (
+	"context"
+	"strconv"
+
+	"go.opencensus.io/stats"
+	"go.opencensus.io/stats/view"
+	"go.opencensus.io/tag"
+	"google.golang.org/grpc/codes"
+)
+
+const (
+	typeURL      = "typeURL"
+	errorCode    = "code"
+	errorStr     = "error"
+	connectionID = "connectionID"
+)
+
+type Reporter interface {
+	SetClientsTotal(clients int64)
+	RecordSendError(err error, code codes.Code)
+	RecordRecvError(err error, code codes.Code)
+	RecordRequestSize(typeURL string, connectionID int64, size int)
+	RecordRequestAck(typeURL string, connectionID int64)
+	RecordRequestNack(typeURL string, connectionID int64)
+}
+
+type StatsContext struct {
+	clientsTotal      *stats.Int64Measure
+	requestSizesBytes *stats.Int64Measure
+	requestAcksTotal  *stats.Int64Measure
+	requestNacksTotal *stats.Int64Measure
+	sendFailuresTotal *stats.Int64Measure
+	recvFailuresTotal *stats.Int64Measure
+}
+
+func (s *StatsContext) SetClientsTotal(clients int64) {
+	stats.Record(context.Background(), s.clientsTotal.M(clients))
+}
+
+func (s *StatsContext) recordError(err error, code codes.Code, stat *stats.Int64Measure) {
+	ctx, ctxErr := tag.New(context.Background(),
+		tag.Insert(ErrorTag, err.Error()),
+		tag.Insert(ErrorCodeTag, strconv.FormatUint(uint64(code), 10)))
+	if ctxErr != nil {
+		scope.Errorf("MCP: error creating monitoring context. %", ctxErr)
+		return
+	}
+	stats.Record(ctx, stat.M(1))
+}
+
+func (s *StatsContext) RecordSendError(err error, code codes.Code) {
+	s.recordError(err, code, s.sendFailuresTotal)
+}
+
+func (s *StatsContext) RecordRecvError(err error, code codes.Code) {
+	s.recordError(err, code, s.recvFailuresTotal)
+}
+
+func (s *StatsContext) RecordRequestSize(typeURL string, connectionID int64, size int) {
+	ctx, ctxErr := tag.New(context.Background(),
+		tag.Insert(TypeURLTag, typeURL),
+		tag.Insert(ConnectionIDTag, strconv.FormatInt(connectionID, 10)))
+	if ctxErr != nil {
+		scope.Errorf("MCP: error creating monitoring context. %", ctxErr)
+		return
+	}
+	stats.Record(ctx, s.requestSizesBytes.M(int64(size)))
+}
+
+func (s *StatsContext) RecordRequestAck(typeURL string, connectionID int64) {
+	ctx, ctxErr := tag.New(context.Background(),
+		tag.Insert(TypeURLTag, typeURL),
+		tag.Insert(ConnectionIDTag, strconv.FormatInt(connectionID, 10)))
+	if ctxErr != nil {
+		scope.Errorf("MCP: error creating monitoring context. %", ctxErr)
+		return
+	}
+	stats.Record(ctx, s.requestAcksTotal.M(1))
+}
+
+func (s *StatsContext) RecordRequestNack(typeURL string, connectionID int64) {
+	ctx, ctxErr := tag.New(context.Background(),
+		tag.Insert(TypeURLTag, typeURL),
+		tag.Insert(ConnectionIDTag, strconv.FormatInt(connectionID, 10)))
+	if ctxErr != nil {
+		scope.Errorf("MCP: error creating monitoring context. %", ctxErr)
+		return
+	}
+	stats.Record(ctx, s.requestNacksTotal.M(1))
+}
+
+func NewReporter(prefix string) *StatsContext {
+	ctx := &StatsContext{
```

Comment:
This function assumes the prefix includes a trailing slash. This is potentially error prone if the user forgets to include the prefix. Can we normalize the prefix, i.e. add slash if its not specified?